File: android/app/src/main/java/com/x1f4r/legioncontrol/net/AppUpdates.kt
package com.x1f4r.legioncontrol.net

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import androidx.core.content.FileProvider
import com.x1f4r.legioncontrol.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest

/**
 * Keeping the app itself up to date, from the GitHub releases of whichever repository the
 * configuration names.
 *
 * A size and a version string detect a packaging mistake and prove nothing about who built the file. Android's own
 * installer will refuse an APK signed by a different key, which is a real backstop, but by then the
 * app has already downloaded an attacker's file, told the user it was an update, and handed it to
 * the system installer with a name it made up. Everything below happens before that point.
 *
 * The order is fixed and each step depends on the one before:
 *
 * 1. Read the release. Take the manifest and its detached signature by exact name.
 * 2. Verify the signature against the key pinned in [com.x1f4r.legioncontrol.data.Trust]. Nothing is
 *    read out of the manifest before this succeeds.
 * 3. Take the artefact whose name is exactly `Legion-Control-android-arm64.apk`, never the first
 *    thing ending in `.apk`.
 * 4. Download it, and check its size and sha256 against the signed manifest.
 * 5. Look inside the archive: the package name, the version name and the signing certificate all
 *    have to match the app that is running.
 * 6. Only then offer it to the system installer.
 *
 * Deliberately built on HttpURLConnection rather than pulling in an http client: three requests, no
 * streaming API worth the dependency, and the rest of this project has stayed free of them.
 */
object AppUpdates {

    private fun releasesUrl(repo: String) = "https://api.github.com/repos/$repo/releases/latest"

    private val json = Json { ignoreUnknownKeys = true }

    data class Release(
        val version: String,
        val notes: String,
        val assetUrl: String,
        val assetName: String,
        val sizeBytes: Long,
        /** The signed manifest this release was verified against. */
        val manifest: VerifiedManifest,
    ) {
        /** What the release says the agent version is, for the agent install action. */
        val agentVersion: String get() = manifest.agentVersion
    }

    sealed interface Check {
        /**
         * Nothing to install. [noReleaseYet] when the repository has published none at all, which
         * is a different sentence on screen and the same outcome: there is nothing to press.
         */
        data class UpToDate(val noReleaseYet: Boolean = false) : Check
        data class Available(val release: Release) : Check
        data class Failed(val reason: String) : Check

        /**
         * A release exists and this build cannot verify it against itself.
         *
         * A debug build is signed with the debug key and carries a different package name, so a
         * release APK is legitimately not installable over it. Saying so is better than a signature
         * error from the system installer that reads like the release is broken.
         */
        data class NotForThisBuild(val reason: String) : Check
    }

    /**
     * What an HTTP status from the releases endpoint means on its own. Null when the body decides.
     *
     * A 404 is not a failure worth raising. GitHub answers it both for a repository with no
     * releases and for one it will not show to an anonymous request, and in both cases the honest
     * thing to say is that there is no release to install, not that something went wrong.
     */
    internal fun outcomeOf(responseCode: Int): Check? = when (responseCode) {
        200 -> null
        404 -> Check.UpToDate(noReleaseYet = true)
        else -> Check.Failed("GitHub answered $responseCode")
    }

    /** Ask GitHub what the newest release is, verify its manifest, and compare it to what is here. */
    suspend fun check(repo: String): Check = withContext(Dispatchers.IO) {
        if (BuildConfig.DEBUG) {
            return@withContext Check.NotForThisBuild(
                "This is a debug build. It is signed with the debug key and installs under a " +
                    "different package name, so a published release cannot replace it.",
            )
        }
        try {
            val connection = open(releasesUrl(repo), "application/vnd.github+json")
            outcomeOf(connection.responseCode)?.let { return@withContext it }

            val body = connection.inputStream.bufferedReader().use { it.readText() }
            val root = json.parseToJsonElement(body).jsonObject
            val tag = root["tag_name"]?.jsonPrimitive?.content.orEmpty()
            val notes = root["body"]?.jsonPrimitive?.content.orEmpty()
            val assets = root["assets"]?.jsonArray?.map { it.jsonObject }.orEmpty()

            fun assetUrl(name: String): String? = assets
                .firstOrNull { it["name"]?.jsonPrimitive?.content == name }
                ?.get("url")?.jsonPrimitive?.content

            fun assetSize(name: String): Long = assets
                .firstOrNull { it["name"]?.jsonPrimitive?.content == name }
                ?.get("size")?.jsonPrimitive?.content?.toLongOrNull() ?: 0L

            val latest = tag.removePrefix("v")
            if (!isNewer(latest, BuildConfig.VERSION_NAME)) return@withContext Check.UpToDate()

            val manifestBytes = assetUrl(ReleaseAssets.MANIFEST)?.let { fetch(it) }
            val signature = assetUrl(ReleaseAssets.MANIFEST_SIGNATURE)
                ?.let { fetch(it) }
                ?.toString(Charsets.UTF_8)

            val verified = ReleaseVerification.verify(manifestBytes, signature, latest)
                .getOrElse { failure ->
                    return@withContext Check.Failed(
                        (failure as? ManifestVerificationFailed)?.problem?.sentence
                            ?: failure.message
                            ?: "the release manifest could not be verified",
                    )
                }

            if (verified.artifact(ReleaseAssets.ANDROID_APK) == null) {
                return@withContext Check.Failed(
                    ManifestProblem.MissingArtifact(ReleaseAssets.ANDROID_APK).sentence,
                )
            }
            val url = assetUrl(ReleaseAssets.ANDROID_APK) ?: return@withContext Check.Failed(
                ManifestProblem.ArtifactMissingFromRelease(ReleaseAssets.ANDROID_APK).sentence,
            )

            Check.Available(
                Release(
                    version = latest,
                    notes = notes.lineSequence().take(6).joinToString("\n").trim(),
                    assetUrl = url,
                    assetName = ReleaseAssets.ANDROID_APK,
                    sizeBytes = assetSize(ReleaseAssets.ANDROID_APK),
                    manifest = verified,
                ),
            )
        } catch (e: Exception) {
            Check.Failed(e.message ?: e.javaClass.simpleName)
        }
    }

    /**
     * Numeric field by field, so 1.0.10 is correctly newer than 1.0.9. A string comparison gets that
     * backwards, and it is exactly the version where it would start to matter.
     */
    internal fun isNewer(candidate: String, installed: String): Boolean {
        fun parts(v: String) = v.substringBefore('-').split('.').map { it.toIntOrNull() ?: 0 }
        val a = parts(candidate)
        val b = parts(installed)
        for (i in 0 until maxOf(a.size, b.size)) {
            val x = a.getOrElse(i) { 0 }
            val y = b.getOrElse(i) { 0 }
            if (x != y) return x > y
        }
        return false
    }

    /** What a download ended in. A file only ever appears on success. */
    sealed interface Download {
        data class Ready(val file: File) : Download
        data class Refused(val problem: ManifestProblem) : Download
        data class Failed(val reason: String) : Download
    }

    /**
     * Download into app private storage, then check it against the signed manifest and against this
     * app's own identity. onProgress gets 0..100, or -1 when the size is unknown.
     *
     * The file is written to a staging name and only renamed into place once every check has
     * passed, so there is never a moment where a half-checked APK sits at the path the install
     * action reads. A refused download leaves nothing behind at all.
     */
    suspend fun download(
        context: Context,
        release: Release,
        onProgress: (Int) -> Unit,
    ): Download = withContext(Dispatchers.IO) {
        try {
            val dir = File(context.cacheDir, "updates").apply { mkdirs() }
            // One file, reused. A cache full of old apks is not worth keeping around.
            dir.listFiles()?.forEach { it.delete() }
            val staging = File(dir, "${release.assetName}.part")
            val target = File(dir, release.assetName)

            val connection = open(release.assetUrl, "application/octet-stream").apply {
                instanceFollowRedirects = true
                readTimeout = 60_000
            }
            if (connection.responseCode != 200) {
                return@withContext Download.Failed("download failed: HTTP ${connection.responseCode}")
            }

            val expected = release.manifest.artifact(release.assetName)?.size
                ?.takeIf { it > 0 }
                ?: release.sizeBytes.takeIf { it > 0 }
                ?: connection.contentLengthLong

            // Hashed while it is written rather than by reading the file again afterwards. It is one
            // pass over the same bytes and it means the hash is of what was received, not of what is
            // on disk a moment later.
            val digest = MessageDigest.getInstance("SHA-256")
            var written = 0L
            connection.inputStream.use { input ->
                staging.outputStream().use { output ->
                    val buffer = ByteArray(64 * 1024)
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) break
                        output.write(buffer, 0, read)
                        digest.update(buffer, 0, read)
                        written += read
                        onProgress(if (expected > 0) ((written * 100) / expected).toInt() else -1)
                    }
                }
            }

            val artifact = release.manifest.artifact(release.assetName)
            if (artifact == null) {
                staging.delete()
                return@withContext Download.Refused(
                    ManifestProblem.MissingArtifact(release.assetName),
                )
            }
            if (artifact.size > 0 && written != artifact.size) {
                staging.delete()
                return@withContext Download.Refused(
                    ManifestProblem.WrongSize(release.assetName, artifact.size, written),
                )
            }
            val actual = digest.digest().joinToString("") { "%02x".format(it) }
            if (!actual.equals(artifact.sha256, ignoreCase = true)) {
                staging.delete()
                return@withContext Download.Refused(ManifestProblem.WrongHash(release.assetName))
            }

            // The archive itself, before it is ever named as something to install. Package, version
            // and signing certificate, all against the app that is running.
            ApkInspection.check(context, staging, expectedVersion = release.version)?.let { problem ->
                staging.delete()
                return@withContext Download.Failed(problem)
            }

            if (!staging.renameTo(target)) {
                staging.copyTo(target, overwrite = true)
                staging.delete()
            }
            Download.Ready(target)
        } catch (e: Exception) {
            Download.Failed(e.message ?: e.javaClass.simpleName)
        }
    }

    /** True when Android will let us hand it an apk at all. */
    fun canInstall(context: Context): Boolean = context.packageManager.canRequestPackageInstalls()

    /** Send the user to the one setting that has to be on before an install can be offered. */
    fun installPermissionIntent(context: Context): Intent =
        Intent(android.provider.Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES)
            .setData(Uri.parse("package:${context.packageName}"))

    /**
     * Hand the apk to the system installer. It shows its own confirmation, which is the right place
     * for that decision to be made, so this does not ask again first.
     *
     * By the time anything reaches here the file has been through the signed manifest and the
     * archive inspection. The installer's own signature check is the last of four, not the first.
     */
    fun install(context: Context, apk: File) {
        val uri = FileProvider.getUriForFile(context, "${context.packageName}.updates", apk)
        val intent = Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, "application/vnd.android.package-archive")
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        context.startActivity(intent)
    }

    private fun open(url: String, accept: String): HttpURLConnection =
        (URL(url).openConnection() as HttpURLConnection).apply {
            setRequestProperty("Accept", accept)
            setRequestProperty("User-Agent", "legion-control")
            connectTimeout = 10_000
            readTimeout = 15_000
        }

    /** A small asset, whole. Used for the manifest and its signature, which are both tiny. */
    private fun fetch(url: String): ByteArray? {
        val connection = open(url, "application/octet-stream").apply {
            instanceFollowRedirects = true
        }
        if (connection.responseCode != 200) return null
        return connection.inputStream.use { it.readBytes() }
    }
}

/**
 * What is actually inside a downloaded APK.
 *
 * The signed manifest proves the bytes are the ones the release was built from. This proves the
 * bytes are an update to *this* app: the same package, a version that matches what the release
 * claims, and the same signing certificate. Android would refuse a mismatched certificate at install
 * time anyway, and finding out here means the user is told why instead of being handed an installer
 * that fails with a message about certificates.
 */
object ApkInspection {

    /** Null when the archive is a genuine update to this app, or the sentence saying why not. */
    fun check(context: Context, apk: File, expectedVersion: String): String? {
        val flags = PackageManager.GET_SIGNING_CERTIFICATES
        val archive = context.packageManager.getPackageArchiveInfo(apk.absolutePath, flags)
            ?: return "That file is not an Android package this phone can read."

        if (archive.packageName != context.packageName) {
            return "That package is ${archive.packageName} and this app is ${context.packageName}."
        }

        val version = archive.versionName
        if (version != null && version != expectedVersion) {
            return "That package says it is version $version and the release says $expectedVersion."
        }

        val installed = signingDigests(context, context.packageName)
        val offered = signingDigests(archive)
        if (installed.isEmpty() || offered.isEmpty()) {
            return "The signing certificate of that package could not be read, so it was not offered."
        }
        if (installed.intersect(offered).isEmpty()) {
            return "That package is signed by a different key than the app that is running. " +
                "Nothing was installed."
        }
        return null
    }

    private fun signingDigests(context: Context, packageName: String): Set<String> = runCatching {
        val info = context.packageManager.getPackageInfo(
            packageName,
            PackageManager.GET_SIGNING_CERTIFICATES,
        )
        signingDigests(info)
    }.getOrDefault(emptySet())

    @Suppress("DEPRECATION")
    private fun signingDigests(info: android.content.pm.PackageInfo): Set<String> {
        val signatures = info.signingInfo?.let { signing ->
            if (signing.hasMultipleSigners()) signing.apkContentsSigners else signing.signingCertificateHistory
        } ?: return emptySet()
        return signatures.orEmpty().map { signature ->
            MessageDigest.getInstance("SHA-256").digest(signature.toByteArray())
                .joinToString("") { "%02x".format(it) }
        }.toSet()
    }
}

File: mac/Sources/LegionControl/AppUpdates.swift
import Foundation

/// Keeping Legion Control itself up to date, from the GitHub releases of whichever repository the
/// configuration names.
///
/// The same releases the phone reads, so the same conventions: one `vX.Y.Z` tag per release, the
/// version taken off the tag, and one archive of the app attached to it. What is different is the
/// install. Android hands an apk to the system installer and the system takes it from there; a Mac
/// app has to replace itself, while it is running, and then start the replacement.
///
/// That is why the order below is download, unpack, check, and only then swap. Everything that can
/// still go wrong happens in a temporary directory, and the one irreversible step is last. A build
/// that turns out to be truncated, unreadable or the wrong version must never have touched
/// /Applications, because a half replaced bundle is an app that will not start again and cannot fix
/// itself.
///
/// Built on URLSession and the command line tools that ship with the system, for the same reason the
/// rest of this project has no dependencies: two requests and four processes are not worth one.
enum AppUpdates {

    /// What a release of the Mac app is shipped as: `ditto -c -k --keepParent` of the bundle, which
    /// is what mac/build.sh produces and what scripts/release.sh attaches.
    static let assetSuffix = ".zip"

    /// The newest release, once it is known to be newer than what is running.
    struct Release: Sendable, Equatable {
        var version: String
        /// The first paragraph of the release body, or empty. Nobody reads a changelog in a
        /// settings row; the first paragraph is the sentence that says what changed.
        var notes: String
        var assetURL: URL
        var sizeBytes: Int64
    }

    enum Check: Sendable, Equatable {
        /// Nothing to install. The note carries the quiet reason when there is one worth saying,
        /// which today means a repository that has never cut a release at all.
        case upToDate(note: String?)
        case available(Release)
        case failed(reason: String)
    }

    // MARK: - Asking

    /// Ask GitHub what the newest release is, and whether it is newer than what is running.
    ///
    /// No credentials, and no way to pass any: this app only ever reads a public repository's
    /// releases. A fork that is private simply has nothing here to read, which the 404 below says
    /// as plainly as it can.
    static func check(repo: String, installedVersion: String) async -> Check {
        guard let url = URL(string: "https://api.github.com/repos/\(repo)/releases/latest") else {
            return .failed(reason: "\"\(repo)\" is not a repository name this can ask about.")
        }

        var request = URLRequest(url: url, timeoutInterval: 10)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("legion-control", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            // 404 is what a public repository that has never cut a release answers, and it is also
            // what a private one answers to anyone who cannot see it. Neither is a failure worth
            // colouring: there is nothing to install, and saying so quietly is the honest reading.
            if code == 404 {
                return .upToDate(note: "\(repo) has no release yet, so there is nothing newer to install.")
            }
            guard code == 200 else { return .failed(reason: "GitHub answered \(code).") }
            return read(releaseJSON: data, installedVersion: installedVersion)
        } catch {
            return .failed(reason: error.localizedDescription)
        }
    }

    /// Everything about a reply that does not need the network: the tag, the comparison and the
    /// asset. Kept apart from the request above so it can be exercised on its own, because this is
    /// the part that has an answer that can be wrong rather than merely absent.
    static func read(releaseJSON data: Data, installedVersion: String) -> Check {
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return .failed(reason: "GitHub's answer could not be read.")
        }

        let latest = version(fromTag: root["tag_name"] as? String ?? "")
        guard !latest.isEmpty else {
            return .failed(reason: "The newest release has no tag to take a version from.")
        }
        guard isNewer(latest, than: installedVersion) else { return .upToDate(note: nil) }

        let assets = root["assets"] as? [[String: Any]] ?? []
        guard let asset = assets.first(where: { ($0["name"] as? String)?.hasSuffix(assetSuffix) == true }),
              // The asset API url rather than browser_download_url: the same one line of code then
              // works for a public repository and, with a header added, for a private fork.
              let assetURL = (asset["url"] as? String).flatMap(URL.init(string:))
        else {
            return .failed(reason: "Release \(latest) has no \(assetSuffix) attached, so there is nothing to install.")
        }

        return .available(Release(
            version: latest,
            notes: firstParagraph(of: root["body"] as? String ?? ""),
            assetURL: assetURL,
            sizeBytes: (asset["size"] as? NSNumber)?.int64Value ?? 0
        ))
    }

    /// The tag with its leading `v` taken off, which is the whole of the release naming convention.
    static func version(fromTag tag: String) -> String {
        let trimmed = tag.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.hasPrefix("v") else { return trimmed }
        return String(trimmed.dropFirst())
    }

    /// Numeric field by field, so 1.0.10 is correctly newer than 1.0.9. A string comparison gets
    /// that backwards, and it is exactly the version where it would start to matter.
    static func isNewer(_ candidate: String, than installed: String) -> Bool {
        func fields(_ value: String) -> [Int] {
            value.split(separator: "-", maxSplits: 1)[0]
                .split(separator: ".")
                .map { Int($0) ?? 0 }
        }
        let a = fields(candidate)
        let b = fields(installed)
        for index in 0..<max(a.count, b.count) {
            let left = index < a.count ? a[index] : 0
            let right = index < b.count ? b[index] : 0
            if left != right { return left > right }
        }
        return false
    }

    /// The first paragraph of a release body, as one line. Release notes are prose separated by
    /// blank lines, and the row this lands in is one sentence wide.
    static func firstParagraph(of body: String) -> String {
        let normalised = body.replacingOccurrences(of: "\r\n", with: "\n")
        let paragraph = normalised
            .components(separatedBy: "\n\n")
            .first { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty } ?? ""
        return paragraph
            .split(separator: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // MARK: - Installing

    /// Why an install stopped, as a sentence to put on screen. Every one of these names the thing
    /// that went wrong rather than the step that noticed it.
    struct InstallFailure: Error, Sendable {
        var message: String

        init(_ message: String) { self.message = message }
    }

    /// A downloaded archive and the directory holding it. The caller owns both and hands them back
    /// to `discard` when it is done, whether the install went through or not.
    struct Download: Sendable {
        var directory: URL
        var archive: URL
    }

    /// Fetch the asset into a directory of its own. Nothing outside that directory is touched here.
    static func download(_ release: Release) async throws -> Download {
        let directory = URL(fileURLWithPath: NSTemporaryDirectory())
            .appending(path: "legion-control-update-\(UUID().uuidString)")
        do {
            try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        } catch {
            throw InstallFailure("A temporary directory for the download could not be made. \(error.localizedDescription)")
        }

        var request = URLRequest(url: release.assetURL, timeoutInterval: 60)
        request.setValue("application/octet-stream", forHTTPHeaderField: "Accept")
        request.setValue("legion-control", forHTTPHeaderField: "User-Agent")

        let archive = directory.appending(path: "Legion-Control.zip")
        do {
            // A download task rather than a data task: the archive is tens of megabytes and there is
            // no reason for all of it to sit in memory on its way to a file.
            let (temporary, response) = try await URLSession.shared.download(for: request)
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard code == 200 else {
                try? FileManager.default.removeItem(at: directory)
                throw InstallFailure("The download failed. GitHub answered \(code).")
            }
            try FileManager.default.moveItem(at: temporary, to: archive)
        } catch let failure as InstallFailure {
            throw failure
        } catch {
            try? FileManager.default.removeItem(at: directory)
            throw InstallFailure("The download failed. \(error.localizedDescription)")
        }

        // A truncated archive unpacks into something that looks almost right, so the size is checked
        // here where it is still only a file in /tmp.
        let written = (try? FileManager.default.attributesOfItem(atPath: archive.path)[.size] as? NSNumber)??.int64Value ?? 0
        if release.sizeBytes > 0, written != release.sizeBytes {
            try? FileManager.default.removeItem(at: directory)
            throw InstallFailure("The download stopped at \(written) of \(release.sizeBytes) bytes.")
        }

        return Download(directory: directory, archive: archive)
    }

    /// Unpack, check what came out, and only then replace the bundle this app is running out of.
    static func apply(_ download: Download, version: String, replacing bundleURL: URL) async throws {
        guard bundleURL.pathExtension == "app" else {
            throw InstallFailure("Legion Control is not running out of an app bundle, so there is nothing to replace. Build and install it with mac/build.sh instead.")
        }

        let unpacked = download.directory.appending(path: "unpacked")
        let extraction = await Shell.run(
            executable: "/usr/bin/ditto",
            arguments: ["-x", "-k", download.archive.path, unpacked.path],
            timeout: 180
        )
        guard extraction.succeeded else {
            throw InstallFailure("The download could not be unpacked. \(extraction.failureText)")
        }

        let replacement = try bundle(in: unpacked)
        try await verify(replacement, is: version)

        // The build is ad hoc signed and arrived over the network, so it carries the quarantine flag
        // and Gatekeeper would refuse to launch it with no way to say yes to something that has no
        // developer identity to trust. Stripped here, before it becomes the app in /Applications.
        _ = await Shell.run(
            executable: "/usr/bin/xattr",
            arguments: ["-dr", "com.apple.quarantine", replacement.path],
            timeout: 60
        )

        // rsync rather than a move: /Applications/Legion Control.app is the bundle this process is
        // running out of, and replacing the directory wholesale would pull it out from under a
        // running app. rsync writes each file beside its target and renames it over the top, so the
        // running executable keeps the inode it started on and nothing is missing in between.
        let swap = await Shell.run(
            executable: "/usr/bin/rsync",
            arguments: ["-a", "--delete", replacement.path + "/", bundleURL.path + "/"],
            timeout: 180
        )
        guard swap.succeeded else {
            throw InstallFailure("The new build could not be put in place. \(swap.failureText)")
        }
    }

    /// Throws the download away. Safe to call twice, and called on the way out of a failure as well
    /// as a success: a few tens of megabytes in /tmp is not something to leave behind either way.
    static func discard(_ download: Download) {
        try? FileManager.default.removeItem(at: download.directory)
    }

    /// Start the replacement a moment from now, from a process that is not this one.
    ///
    /// `open` on a bundle that is mid-quit gets the app that is already running rather than a new
    /// one, so the launch is handed to a shell that waits a second first and outlives us. Nothing
    /// waits on it: by the time it runs, this process is meant to be gone.
    static func relaunch(_ bundleURL: URL) {
        let quoted = bundleURL.path
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/sh")
        process.arguments = ["-c", "sleep 1; open -n \"\(quoted)\""]
        process.standardInput = FileHandle.nullDevice
        process.standardOutput = FileHandle.nullDevice
        process.standardError = FileHandle.nullDevice
        try? process.run()
    }

    // MARK: - Checking what was unpacked

    /// The one app bundle the archive held. Anything else means the asset is not what this expects,
    /// and guessing at that point would be guessing about what to overwrite /Applications with.
    private static func bundle(in directory: URL) throws -> URL {
        let contents = (try? FileManager.default.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: nil,
            options: [.skipsHiddenFiles]
        )) ?? []
        let bundles = contents.filter { $0.pathExtension == "app" }
        guard bundles.count == 1, let only = bundles.first else {
            throw InstallFailure("The archive did not hold exactly one app bundle, so there is nothing safe to install from it.")
        }
        return only
    }

    /// That the thing about to be copied over /Applications really is this app, and really is the
    /// version the release said it was. Both checks are cheap and both have caught the two ways a
    /// release goes wrong: an asset built from the wrong tree, and a tag bumped without a rebuild.
    private static func verify(_ bundle: URL, is version: String) async throws {
        let executable = bundle.appending(path: "Contents/MacOS/LegionControl")
        guard FileManager.default.isExecutableFile(atPath: executable.path) else {
            throw InstallFailure("The downloaded bundle has no Contents/MacOS/LegionControl in it, so it is not this app.")
        }

        let plist = bundle.appending(path: "Contents/Info.plist")
        let read = await Shell.run(
            executable: "/usr/libexec/PlistBuddy",
            arguments: ["-c", "Print :CFBundleShortVersionString", plist.path],
            timeout: 30
        )
        guard read.succeeded else {
            throw InstallFailure("The downloaded bundle's version could not be read. \(read.failureText)")
        }
        let found = read.standardOutput.trimmingCharacters(in: .whitespacesAndNewlines)
        guard found == version else {
            throw InstallFailure("The release says \(version) but the bundle in it says \(found), so it was not installed.")
        }
    }
}
